flipperdevices/flipperzero-firmware
# [FL-3893] JS modules

# What's new
  - JS modules: `event_loop`, `gpio`,`gui` and `storage`
    - JS `gui` submodules: `gui/loading`, `gui/submenu`, `gui/empty_screen`, `gui/text_input`, `gui/text_box`, `gui/dialog`
  - Documentation for these modules
  - TS type declarations for these modules
  - JS example scripts for these modules (except `storage`)
  - Some tests for some of the modules
  - Removed legacy JS modules: `submenu` (superseded by `gui/submenu`), `textbox` (superseded by `gui/text_box`), `dialog` (superseded by `gui/dialog`)
  - `TextInputModel` field: `minimum_length`
  - SDK API functions: `text_input_set_minimum_length` , `furi_event_loop_is_subscribed`
  - Static inline convenience function: `furi_event_loop_maybe_unsubscribe`

# Verification 
  - Run unit tests
  - For modules that require access to the outside world (`gpio`, `gui`):
    - Build and flash the firmware
    - Run `gpio` and `gui` JS example scripts
    - Verify that the examples work as intended

# Remarks
## Performance
The GUI example is quite heavy, requiring about 32 KiB of heap space with all the GUI modules loaded. The script starts up slowly compared to other scripts (~500ms), but runs just fine.

## Readability
The later the date of creation of a module, the more its implementation looks like it's written in a DSL. Without all these macros the code would just become 80% noise, and thus a huge pain in the butt to write and maintain.

# Checklist (For Reviewer)
